src/webgpu/api/operation/rendering/stencil.spec.ts

```diff
@@ -0,0 +1,230 @@
+
+const kBaseColor = [0xff, 0xff, 0xff, 0xff];
+const kStencilColor = [0x00, 0xff, 0x00, 0xff];
+
+class StencilTest extends GPUTest {
+  checkStencilCompareFunction(
+    compareFunction: GPUCompareFunction,
+    stencilRefValue: number,
+    expectStencilColor: boolean
+  ) {
+    const depthStencilFormat: GPUTextureFormat = 'depth24plus-stencil8';
+
+    const baseStencilState = {
+      compare: 'always',
+      failOp: 'keep',
+      passOp: 'replace',
+    } as const;
+
+    const stencilState = {
+      compare: compareFunction,
+      failOp: 'keep',
+      passOp: 'keep',
+    } as const;
+
+    const baseState = {
+      format: depthStencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always',
+      stencilFront: baseStencilState,
+      stencilBack: baseStencilState,
+    } as const;
+
+    const state = {
+      format: depthStencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always',
+      stencilFront: stencilState,
+      stencilBack: stencilState,
+    } as const;
+
+    this.runStencilStateTest(
+      baseState,
+      state,
+      stencilRefValue,
+      expectStencilColor ? kStencilColor : kBaseColor
+    );
+  }
+
+  runStencilStateTest(
+    baseState: GPUDepthStencilState,
+    state: GPUDepthStencilState,
+    stencilRefValue: number,
+    stencilColor: number[]
+  ) {
+    const renderTargetFormat = 'rgba8unorm';
+    const renderTarget = this.device.createTexture({
+      format: renderTargetFormat,
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const depthStencilFormat: GPUTextureFormat = 'depth24plus-stencil8';
+    const depthTexture = this.device.createTexture({
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      format: depthStencilFormat,
+      sampleCount: 1,
+      mipLevelCount: 1,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_DST,
+    });
+
+    const depthStencilAttachment: GPURenderPassDepthStencilAttachment = {
+      view: depthTexture.createView(),
+      depthLoadOp: 'load',
+      depthStoreOp: 'store',
+      stencilLoadOp: 'load',
+      stencilStoreOp: 'store',
+    };
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: renderTarget.createView(),
+          storeOp: 'store',
+          loadOp: 'load',
+        },
+      ],
+      depthStencilAttachment,
+    });
+
+    // Draw the base triangle with stencil reference 1.
+    // This clears the stencil buffer to 1.
+    {
+      const testPipeline = this.createRenderPipelineForTest(baseState);
+      pass.setPipeline(testPipeline);
+      pass.setStencilReference(1);
+      pass.setBindGroup(
+        0,
+        this.createBindGroupForTest(
+          testPipeline.getBindGroupLayout(0),
+          new Float32Array(kBaseColor)
```

Comment:
The members of this Float32Array are initialized to 0xff (255.0) and 0x00 (0.0). This technically works, they get clamped to the 0.0-1.0 range, but then the same values are used to initialize a Uint8Array later resulting in the unorm values 0xff (1.0) and 0x00 (0.0) which seems accidental.

Reply:
Ah, thank you for pointing it out. Then, how about using textureContentIsOKByT2B to check the result instead of expectSinglePixelIn2DTexture?

Reply:
Sounds good.